internal/statemachine/helper.go

```diff
@@ -793,6 +793,11 @@ func manualCopyFile(copyFileInterfaces interface{}, targetDir string, debug bool
 		if debug {
 			fmt.Printf("Copying file \"%s\" to \"%s\"", copyFile.Source, dest)
 		}
+		absDest, _ := filepath.Abs(dest)
```

Comment:
Slightly different approach, but maybe safer -- call filepath.abs or filepath.clean on copyFile.Dest. perhaps, by default (I'd need to look at CopyFileType and see if it could be a part of the .Dest call?)

That way before the join, we have a clean target. I'd need to check targetDir code again to see if there's escape there(and I'm mobile in Greece do not let my partner know...)

Reply:
It might be a bit too late here, but why would it be safer? Also, without joining it with targetDir, basically a `filepath.Abs()` call would assume the path to be relative to the current working directory, which is not true and can potentially just add to the confusion (since paths given here don't have to be absolute). After joining with targetDir we sanitize the final path which will be used, which seems the safest - since *this* will be the path that will be used ultimately, no?

Reply:
Like i said, i need to re-orient myself on how the chroot is created. From my perspective it's about where a user may put relative path moves, and the check. 

I'm not sure its safer, but from a general standpoint, one often santizes on input, not just execution. Input sanitization provides early feedback and prevents unsafe code to even reach an execution phase.

In the case `Workdir`, i don't see input santization done. If `targetDir` == `Workdir`and those are relative, then they'll fail the `hasPrefix` check later due to the Abs expansion. Thus, don't we need to ensure `targetDir` is expanded and sanitized before the check below in case it is relative? 

Also, is ubuntu-image a confined snap? If so, that puts some good constraints on paths. What i dont want is someone accidentally doing this and overwriting their own `etc` or something due to a bad combo of workdir + dest. And for the requirements to be clearly stated to a user as early as possible. A failure on input validation is nice (especially if it's generated paths from multiple inputs, like this func)

Reply:
Due to the nature of how we prime gadgets, sadly ubuntu-image needs to stay as a classic snap. For now.

As for the other discussions: hmmm, so I might try sanitizing the input to the manual customization functions by simply disallowing relative paths for the destination basically during `parseImageDefinition()` - I'll do that now since I actually like that better conceptually! So thanks for the suggestion!
As for Workdir, I wouldn't worry about that so much. The workdir is either manually provided on ubuntu-image execution or is automatically generated by ubuntu-image - so we either have control over it or we leave it up to the user to do a conscious decision on which directory to use.

Reply:
But wouldn't a relative workdir fail test below, as it's expanded after the var set?

So wouldn't we need to expand early or else 

`ubuntu-image -w ../../build/$DATE` in PWD `/home/$USER/github/my_custom_image` would, after expansion, be

`/home/$USER/build/$DATE`, which would fail `hasPrefix`. 

So we should expand at opt set time